app/models/solid_queue/semaphore.rb

```diff
@@ -41,16 +41,30 @@ def signal
       end
 
       private
+
         attr_accessor :job
 
-        def attempt_creation
+        def attempt_creation_with_insert_on_conflict
+          results = Semaphore.insert({ key: key, value: limit - 1, expires_at: expires_at }, unique_by: :key)
+
+          if results.length.zero?
+            limit == 1 ? false : attempt_decrement
+          else
+            true
+          end
+        end
+
+        def attempt_creation_with_create_and_exception_handling
           Semaphore.create!(key: key, value: limit - 1, expires_at: expires_at)
           true
         rescue ActiveRecord::RecordNotUnique
-          if limit == 1 then false
-          else
-            attempt_decrement
-          end
+          limit == 1 ? false : attempt_decrement
+        end
+
+        if ActiveRecord::Base.connection.adapter_name == "PostgreSQL"
```

Comment:
It seems there's a method that checks directly whether the adapter supports `unique_by` exposed in [Active Record's Abstract adapter](https://api.rubyonrails.org/classes/ActiveRecord/ConnectionAdapters/AbstractAdapter.html#method-i-supports_insert_conflict_target-3F), that we could use simply as 
```ruby
connection.supports_insert_conflict_target?
```
I think this might be more robust than checking the adapter name, because other adapters for PostgreSQL might have a different name, like [the PostGIS adapter](https://github.com/rgeo/activerecord-postgis-adapter).

Reply:
So funny and just goes to show a bit of tunnel vision.  I spent a fair amount of time digging around for an alternative to the original nested transactions methodology and in that digging I came across this method in the Rails adaptors.  

The tunnel vision part was no thinking to use those methods as the much, much better way to test which of the code paths to take.

Thanks for this -- it's a great reminder to slow down and think before coding.